In [63]:
import pickle
import pandas as pd
import os
import glob
from pathlib import Path
import numpy as np

We open the first file of the subset, to check what the HDF5 keys are and then we read each of them.

In [7]:
with pd.HDFStore("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5") as hdf:
    print(hdf.keys())

['/analysis/songs', '/metadata/songs', '/musicbrainz/songs']


In [40]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/analysis/songs")

analysis_sample_rate                         audio_md5  danceability  \
0                 22050  a222795e07cd65b7a530f1346f520649           0.0   

    duration  end_of_fade_in  energy  idx_bars_confidence  idx_bars_start  \
0  218.93179           0.247     0.0                    0               0   

   idx_beats_confidence  idx_beats_start         ...          key  \
0                     0                0         ...            1   

   key_confidence  loudness  mode  mode_confidence  start_of_fade_out   tempo  \
0           0.736   -11.197     0            0.636            218.932  92.198   

   time_signature  time_signature_confidence            track_id  
0               4                      0.778  TRAAAAW128F429D538  

[1 rows x 31 columns]

In [21]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/metadata/songs")

analyzer_version  artist_7digitalid  artist_familiarity  artist_hotttnesss  \
0                              165270            0.581794           0.401998   

            artist_id  artist_latitude  artist_location  artist_longitude  \
0  ARD7TVE1187B99BFB1              NaN  California - LA               NaN   

                            artist_mbid artist_name  artist_playmeid genre  \
0  e77e51a5-4761-45b3-9847-2051f811e366      Casual             4479         

   idx_artist_terms  idx_similar_artists      release  release_7digitalid  \
0                 0                    0  Fear Itself              300848   

   song_hotttnesss             song_id             title  track_7digitalid  
0          0.60212  SOMZWCG12A8C13C480  I Didn't Mean To           3401791

In [22]:
pd.read_hdf("data/MillionSongSubset/data/A/A/A/TRAAAAW128F429D538.h5","/musicbrainz/songs")

idx_artist_mbtags  year
0                  0     0

We only need to extract <tt>tempo</tt> and <tt>song_hotttnesss</tt>, here is an example of how to do that on the subset :

In [66]:
tempo = []
hotness = []

files = glob.glob("data/MillionSongSubset/data/A" + "/[A-Z]/[A-Z]/*")
for f in files:
    tempo.append(pd.read_hdf(f,"/analysis/songs")["tempo"][0])
    hotness.append(pd.read_hdf(f,"/metadata/songs")["song_hotttnesss"][0])

In [67]:
tempo = np.asarray(tempo)
hotness = np.asarray(hotness)
print(tempo)
print(hotness)

[ 124.059   80.084   54.874 ...,   96.064  149.012  162.653]
[ 0.54795294  0.47563847         nan ...,         nan         nan  0.270776  ]


In [71]:
print("Number of tracks =", len(files))
print("with missing hotness values =", np.sum(np.isnan(hotness)))

Number of files = 7620
with missing hotness values = 3268


We already got 3268 unknown hotness values and we only tested on a subset of 7620 song, so we can expect to have that information for only a little over half of our final dataset. Maybe we won't use it.

In [64]:
#loading of the dataframe YOU CAN REMOVE THAT IN THE FINAL NOTEBOOK
covers = pickle.load(open("data/covers.p","rb"))
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB       -1          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6       -1          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1       -1          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74       -1          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99       -1          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438       -1          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E       -1          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2       -1          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946       -1          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128       -1          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4       -1          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE       -1          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029       -1          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595       -1          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833       -1          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A       -1          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7       -1          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D       -1          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD       -1          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6       -1          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343       -1          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3       -1          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164       -1          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C       -1         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920       -1         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220       -1         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743       -1         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900       -1         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7       -1         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE       -1         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9       -1       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF       -1       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C       -1       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

Once all the files are accessible on the cluster, we will have to go through our SHS dataset and get those attributes for each track_id.
We will do so in the following way : (the paths are just examples on the subset)

In [72]:
tempo = []
hotness = []

for track in covers["trackID"]:
    folder1 = track[2]
    folder2 = track[3]
    folder3 = track[4]
    folder_path = "data/MillionSongSubset/data/" + folder1 + "/" + folder2 + "/" + folder3 + "/"
    track_path = folder_path + track + ".h5"
    if Path(track_path).exists(): #to delete later
        tempo.append(pd.read_hdf(track_path,"/analysis/songs")["tempo"][0])
        hotness.append(pd.read_hdf(track_path,"/metadata/songs")["song_hotttnesss"][0])

In [73]:
print(len(tempo))
print(np.sum(~np.isnan(hotness)))

204
128


Unsurprisingly, we only found 204 of those tracks in the subset and 128 of them have a hotness value.

### MILESTONE 3

Getting tempo and popularity 

In [181]:
dataset_path = "data/tempo_and_popularity.csv"
full_df = pd.read_csv(dataset_path)

In [182]:
full_df

track_id  song_hotttnesss    tempo
0      TRCKNGE128F92DA3F3              nan   92.195
1      TRIOPLY128F423CFF3   0.634538103814   88.602
2      TRWNDEU128F9329BF7   0.725213133131  140.600
3      TRYOPHS128F146DEFD   0.661648170034   80.917
4      TRQNZCE128E078A9C0   0.450762241641  120.327
5      TRMBSQR128F92DF66E              nan  103.189
6      TREVWUZ128F4263A9B              nan  113.973
7      TRGYREY128E0791913   0.540009276927  109.904
8      TRBLBSR128F425EBFE              nan   92.555
9      TRLRWJK128F427D602              nan   96.855
10     TRHFEEP12903CD3863              nan  100.113
11     TRWDJFG12903C94C9E              nan  168.998
12     TRLQSWL128F932C80B              0.0  109.548
13     TRHFJNY128F428FC4B              nan  108.374
14     TRMNZOQ128F426254A   0.484562529968   69.319
15     TRYEKWE128F1459C42   0.990837144634  169.616
16     TRPTALT128F426F99F   0.674833876061  152.157
17     TRQPBQV128F423C1E9   0.851363810415  138.338
18     TRTTEXX128F428BE1D              nan  165.623
19     TRYXXWR12903CA7191              nan  133.834
20     TRVXQXV128F427863A              nan  121.932
21     TREVQFK128F4267577   0.672214063234  110.837
22     TRFWVIP128EF3444F3   0.535106024898  107.676
23     TRBZYQA128F9359288              nan   86.319
24     TRTLSVW12903CFB4F5              nan  108.138
25     TRJLBUB128F92E293E   0.215080318509   90.769
26     TRCNOGX128F92E9A8D              nan  144.069
27     TREYSUB128F1485DA2   0.487492443417  143.743
28     TRONGVZ128F9339880              nan  129.208
29     TRIKRKG128F4265ACF   0.599875581301   94.099
...                   ...              ...      ...
16191  TRQUHKZ128F930304E   0.884117263455  104.578
16192  TRDTJZI128EF34E78F   0.449940204043  106.785
16193  TRJUXRE128F92F88D4   0.552705113905  100.986
16194  TRKFNUK128F428A62D   0.760905004619  104.746
16195  TROWHKP12903CCC703   0.783650535257  123.002
16196  TRELXLZ12903CB9DC5              nan  150.633
16197  TRPMGHH128E07917AC   0.828627327993  104.603
16198  TRMTFOJ12903CAFF57              nan   98.177
16199  TRUAXLK12903CD39B5   0.392008771471   97.606
16200  TRNNQSH128F4245865              nan  109.333
16201  TRTWMBY128F9307A6D              nan  100.593
16202  TRITRRU128F423DC76   0.583942265829  142.482
16203  TRRCCUQ128F42449F7   0.212045405484  127.486
16204  TRCEMLE128F4281036   0.743304028064  114.559
16205  TRCGGYF128F9338B93              0.0  163.942
16206  TRAHFYD128F426BBAD   0.509640965552  103.818
16207  TRRUFPY128F4261744              0.0   84.926
16208  TRTFWRF128F92E378E   0.557187767519   61.062
16209  TRBXSHQ128F4278320              0.0   78.955
16210  TRGOMBU128F42895B9              nan  206.595
16211  TRMCRMX128F4267AC3              nan  176.815
16212  TRNBSWN128F1473B31   0.204195042485   88.421
16213  TRWAULS128F1474CBF              nan   77.091
16214  TRMQPSQ128F426C548   0.672177359963  138.849
16215  TRUZAVE128F426E391   0.492452673715  118.558
16216  TRDNIRY128F425A5C8   0.494469709267   89.641
16217  TRSDUWY128F92DD39A              nan  118.440
16218  TRPVDKP128F934956E   0.474882705988  105.135
16219  TREMVLC128F92EFA95   0.439202489315  128.343
16220  TRMTMWV128F92F6F78   0.253834736132  157.587

[16221 rows x 3 columns]

In [183]:
full_df[full_df.song_hotttnesss == "nan"]=np.nan

In [184]:
full_df.count()

track_id           16221
song_hotttnesss    16221
tempo              16221
dtype: int64

In [185]:
#percentage of song_hotttnesss available
100 * ( 1 - full_df.song_hotttnesss.str.contains("nan").sum() / full_df.shape[0] )

62.394426977375005

In [186]:
#percentage of tempo available
100 * ( 1 - (full_df.tempo==0).sum() / full_df.shape[0] )

99.938351519635034

In [187]:
#Number of matches between the full set and the cover set
len(set(full_df.track_id) & set(covers.trackID))

16221

### SUITE DU PIPELINE (TO DELETE)

* détermination de l'original en fonction de l'année pour les unknown :

In [178]:
covers = pickle.load(open("covers_FINAL.p","rb"))
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB        0          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6        0          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1   412972          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74        0          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99        0          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438        0          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E    16660          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2   551633          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946   354066          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128   354067          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4    52010          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE    52011          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029        0          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595        0          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833    45769          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A     9133          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7    59663          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D      575          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD    68273          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6   305654          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343   160334          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3        0          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164    14943          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C    21513         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920    21514         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220    31718         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743    31719         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900    88882         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7    88880         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE    29876         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9   184203       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF        0       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C     9981       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

In [179]:
# general stat
covers[covers["original_id"] == "Unknown"]

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB        0          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6        0          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1   412972          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74        0          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99        0          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438        0          1   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029        0          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595        0          5   
35     TRUATRC12903CAAF21  AR0RCMP1187FB3F427        0         16   
36     TRPJVKE128F428AA27  ARCCN681187B9AFD73        0         16   
45     TRXMDOP12903CBC46A  ARKG7TE1187B9A45D8        0         21   
46     TRKEXQZ128F14959B0  ARJFOC01187FB413A5    60152         21   
47     TRRYIVT128F9336500  AR5HOKG1187FB3873E    60137         21   
52     TRDIRXL128F4222742  ARG81VK1187FB401BF        0         24   
53     TRJZZNO128F92F48DE  ARXFMP01187FB40768        0         24   
54     TRBESBZ12903CA0A95  ARR2BHS1187B9B8FA7        0         25   
55     TROFEEX128F92F190C  AR5JUO91187FB4E90B        0         25   
74     TRSWHFV128F421A882  ARZ7MPB1187B98CA77        0         34   
75     TRHXZRK12903D02C2D  ARH861H1187B9B799E        0         34   
76     TRPKERG128F92CA8F1  AR5US1D1187B9AEB1F        0         35   
77     TRRFVJY128EF3523EC  AR8OVZ01187B9A5E1A        0         35   
86     TRKFOUP128E0786014  ARUJZFJ1187B9B135F        0         40   
87     TRRLKSI128F92D5D37  ARWQ4AM1187B9AD5F8        0         40   
88     TRAOGDO128F42B9A6F  ARPYV1F1187FB444FA        0         41   
89     TRNUMMX128F92D03F4  ARZHQDD1187FB5B871        0         41   
95     TRCLFJM128F92FE654  ARRS1FT1187B9B650A        0         44   
96     TRECEYT12903CB8F86  AR7KA5V1187FB44E6B        0         44   
99     TRYVXMT128F4272422  AREQIT21187B9AEB7F        0         46   
100    TREEKRX128E0790AD3  AR6TNOA1187B9B5C44        0         46   
103    TRGEHDP128F148A1B4  AR9YWMS1187FB43A34   319149         48   
...                   ...                 ...      ...        ...   
18056  TRFZDMW12903CAB4AE  AR772T41187B98E14F   111345       5800   
18057  TRQENZL128F92EBCC9  AR9F71I1187FB46CAF    98385       5800   
18058  TRELMUN128F4221976  AR5XYFL1187FB414C0    98322       5801   
18059  TRHWDRI128F92CE267  AR772T41187B98E14F   146420       5801   
18060  TRXJVYT128F42345CA  ARD0FN11187B9B25CD   101437       5801   
18067  TRRFEGU128F9335471  AREUZEN1187FB48155    85738       5805   
18068  TRJIAUW128F4253585  ARI5E6U1187B9B525A    45026       5805   
18089  TRORIEA128F42B155F  AR4ZNBW1187FB554B8   148752       5814   
18090  TRBZFDK128EF35A246  ARN0YZ71187B9BA57B   109617       5814   
18095  TRCCGSP12903CDC426  AR9VN011187B9ADD25    86489       5817   
18096  TRNSTHZ128F149EBAC  ARB336A1187B9A4583     9880       5817   
18104  TRHBRQQ12903CCA3D6  ARDHAOT11F4C83C0BC    25364       5821   
18105  TRDZUEO128F92E40D1  AREI6PI1187B9911AE    15971       5821   
18106  TRBIKMD128F145C451  ARGGLM71187B9AA273    16179       5822   
18107  TREFDKH128F93135E9  ARBOG0H1187B99F775    91356       5822   
18134  TRYTOUE128F426599B  ARO0QD41187FB3AB0A    15798       5830   
18135  TRQUJCP128F4238206  ARLAUED1187B9ACEAF    16397       5830   
18136  TRPMDDK12903D0D320  AREWAAS1187B98C490        0       5830   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

In [127]:
# The years recovered from the dataset are floats, so we need to strip their decimal part,
# and then we can take the last 4 characters on any date string to get the year.
covers["year"] = covers["date"].replace("\.0","",regex=True).str[-4:]

In [132]:
# In each clique, we rank the versions by year to find the original
covers["rank"] = covers.groupby("clique_id")["year"].rank(method="dense",ascending=False).astype(int)

In [128]:
covers

trackID            artistID  shsPerf  clique_id  \
0      TRGDMZP128F42BC52B  ARB1DDF1187FB4FCFB        0          0   
1      TRCATYW12903D038FE  ARGJEEO1271F573FD6        0          0   
2      TRVMZJZ128F4270CE4  ARY0HTV1187FB4A1B1   412972          0   
3      TRKOINL128F42926C3  ARQ5FSZ1187B98AD74        0          0   
4      TROJZTF128F428B546  ARJN76O1187FB43C99        0          1   
5      TRYQEDQ128F427917C  ARS4KT21187B9B9438        0          1   
6      TRCKNGE128F92DA3F3  AR1CB5G1187B9AFB8E    16660          2   
7      TRIOPLY128F423CFF3  ARKZJ301187FB521B2   551633          2   
8      TRWNDEU128F9329BF7  ARVZWQ31187B9B8946   354066          3   
9      TRYOPHS128F146DEFD  AR6NYHH1187B9BA128   354067          3   
10     TRMBSQR128F92DF66E  ARPI2DX1187FB4CED4    52010          4   
11     TRQNZCE128E078A9C0  ARWILYB1187FB37DFE    52011          4   
12     TRTUJKS128F4262F5F  ARJA1841187FB3A029        0          5   
13     TRLLCAL128F428B903  ARP2RHS1187B991595    73986          5   
14     TREVWUZ128F4263A9B  AR9UYPT1187B9AE833    45769          6   
15     TRGYREY128E0791913  ARQ294N1187FB53D2A        0          6   
16     TRBLBSR128F425EBFE  AR36FFP1187B9926D7    59663          7   
17     TRLRWJK128F427D602  AR6NBDC1187FB4D96D      575          7   
18     TRYEKWE128F1459C42  AR6AK6B1187FB48EDD    68276          8   
19     TRPTALT128F426F99F  ARQRDXP1187B98B9E6   305654          8   
20     TRYXXWR12903CA7191  ARMBTFC1187FB56343   160334          9   
21     TRTTEXX128F428BE1D  ARJVH1S1187FB378C3        0          9   
22     TRQPBQV128F423C1E9  AREH7LP1187B9B7164    14943          9   
23     TRVXQXV128F427863A  ARTO3L81187B9ACA8C    21513         10   
24     TRZJWSX12903CA89C1  ARC1IHZ1187FB4E920    21514         10   
25     TREVQFK128F4267577  ARTQ8QP1187FB3D220    31718         11   
26     TRFWVIP128EF3444F3  ARE66XI1187B9B8743    31719         11   
27     TRXZEBD128F4268BDB  AR33KSV1187FB4B900    88882         12   
28     TRBZYQA128F9359288  AR20SBM1187B9966A7    88880         12   
29     TRTLSVW12903CFB4F5  ARWX5TK1187FB4CDBE    29876         13   
...                   ...                 ...      ...        ...   
18166  TRDTJZI128EF34E78F  AR1BI4V1187B9937A9   184203       5842   
18167  TRJUXRE128F92F88D4  ARHSO041187FB3D06D    33329       5842   
18168  TRKFNUK128F428A62D  ARV3CRH1187B9A1B21    14338       5842   
18169  TROWHKP12903CCC703  ARCGI111187B9B8B99    67211       5843   
18170  TRELXLZ12903CB9DC5  ARYDLXP122988F049E    62368       5843   
18171  TRPMGHH128E07917AC  ARF4L041187FB4D318    99651       5844   
18172  TRMTFOJ12903CAFF57  AR3KVH41187FB3F71F    99652       5844   
18173  TRTWMBY128F9307A6D  ARSX6IX1187B9B640C    99655       5845   
18174  TRNNQSH128F4245865  AR4SL791187B9890BF        0       5845   
18175  TRUAXLK12903CD39B5  AREYJKQ1187B9AEAB5   104087       5845   
18176  TRITRRU128F423DC76  ARJ59K61187FB5B43E    99662       5846   
18177  TRRCCUQ128F42449F7  ARNUU2R1187B9A316C    99675       5846   
18178  TRCEMLE128F4281036  ARJ59K61187FB5B43E    99668       5847   
18179  TRCGGYF128F9338B93  ARYDZKJ1187FB540C6    99680       5847   
18180  TRAHFYD128F426BBAD  AR94ZOI1187FB46BDA   103980       5848   
18181  TRRUFPY128F4261744  ARJWDFW1241B9C69AE   118487       5848   
18182  TRTFWRF128F92E378E  ARX1CYE1187FB3A97B    99730       5849   
18183  TRBXSHQ128F4278320  ARLM6I41187B9BA123    99731       5849   
18184  TRGOMBU128F42895B9  AR14CJ91187FB3A994    99733       5849   
18185  TRMCRMX128F4267AC3  ARSPUHE1187B99DB8A   125607       5850   
18186  TRNBSWN128F1473B31  ARLG9UJ1187B9B6C67    24567       5850   
18187  TRWAULS128F1474CBF  AR7GM0K1187B9B74B4    20304       5851   
18188  TRZVRJL128F92E2984  ARDYR3C1187FB461CE     9982       5851   
18189  TRMQPSQ128F426C548  ARAWPR61187B9B5A4C     9981       5851   
18190  TRSDUWY128F92DD39A  ARR176H1187FB4CDFE    99838       5852   
18191  TRUZAVE128F426E391  AR3DLBB1187B98F3DF   106651       5852   
18192  TRDNIRY12

In [136]:
#number of first version
(covers["rank"] == 1).sum()

6209

In [141]:
#number of cliques
covers.clique_id.max() + 1

5854

In [149]:
unknown_originals = covers[covers["rank"] == 1][covers[covers["rank"] == 1].duplicated(subset="clique_id",keep=False)]

In [171]:
covers[covers["year"].str.contains("nan") & covers["original_id"].str.contains("Unknown")].shape

(154, 13)

In [173]:
covers.drop(covers[covers["year"].str.contains("nan") & covers["original_id"].str.contains("Unknown")].index).shape

(18042, 13)

In [180]:
len(covers[covers.duplicated('shsPerf',keep=False) & covers.shsPerf!=0].sort_values('shsPerf'))

261